### DISCUSSION TITLE
Format shape of virial stresses in training set

### DISCUSSION THREAD

**Discussion by yonghyuklee** [ACEsuit/mace#294] [Q&A] [answered]
Dear developers,

I am encountering difficulties while training MACE on energy, force, and virial stress due to the required format of the stress tensor, which should be (1,3,3) when interpreted by mace_run_train. I have been utilizing my training set in the extended XYZ format, where the virial data is saved as follows, for instance: 
Lattice="9.552033323 0.0 0.0 0.0 9.552033323 0.0 0.0 0.0 9.26248206" Properties=species:S:1:pos:R:3:dft_forces:R:3 dft_energy=-603.38967375 **dft_virial="-81.69521 0.00000 0.00000 0.00000 -81.69521 0.00000 0.00000 0.00000 -102.66851"** pbc="T T T"
While this format works seamlessly with GAP training, it poses an issue with MACE. Instead of being read as a (1,3,3) array, it is interpreted as a (9,) shaped array. It appears that a (1,3,3) array is not accepted as the key value. Could you please assist me in resolving this issue? Is there an example of the training set that I could refer to?

Many thanks!

**Comment by ilyes319** [comment]
Dear @yonghyuklee,

Is your property stress or virials (not normalized by the volume of the cell)? In the case of a virial quantity, you need to input them as a 3x3 matrix. So, in your case, you would need to have  `dft_virial=[[-81.69521 0.00000 0.00000], [0.00000 -81.69521 0.00000], [0.00000 0.00000 -102.66851]]`.

**Reply by yonghyuklee** [reply]
Thank you very much for your reply, and yes the property is virials.

Unfortunately, it brings an error as below:
```
Traceback (most recent call last):
  File "/home/ylee/.conda/envs/torch/bin/mace_run_train", line 8, in <module>
    sys.exit(main())
  File "/home/ylee/.conda/envs/torch/lib/python3.10/site-packages/mace/cli/run_train.py", line 133, in main
    dataset=[
  File "/home/ylee/.conda/envs/torch/lib/python3.10/site-packages/mace/cli/run_train.py", line 134, in <listcomp>
    data.AtomicData.from_config(config, z_table=z_table, cutoff=args.r_max)
  File "/home/ylee/.conda/envs/torch/lib/python3.10/site-packages/mace/data/atomic_data.py", line 178, in from_config
    torch.tensor(config.virials, dtype=torch.get_default_dtype()).unsqueeze(0)
TypeError: new(): invalid data type 'str'
```

The code reads it as a string instead of an array.

**Reply by ilyes319** [reply]
Sorry I edited my answer, could you try without the quote.

**Reply by yonghyuklee** [reply]
It doesn't work as well. When I read it in ASE, atom.info brings as below:
```
In [2]: atoms = ase.io.read("training_set.xyz", ":")

In [3]: atoms[0].info
Out[3]:
{'dft_energy': 17.04652802,
 'dft_virial': '[132.15246 0.00000 0.00000,',
 '0.00000 -0.06439 0.00000,': True,
 '0.00000 0.00000 -0.06299]': True,
 'structure': 'molc',
 'structure_info': 'molc'}
```

**Comment by bernstei** [comment, marked as answer]
The quotes are definitely needed if it's in the xyz comment line. There's also some magic that happens when reading xyz files with 9-long vectors in the comment line being transformed into 3x3 matrices. @jameskermode may remember how that works - it may be some special field names hardwired. If so, the MACE training code might need to reshape the virials field into a matrix if it has length 9.

**Reply by jameskermode** [reply]
In `ase.io.extxyz` the special case conversion from 9-element to 3x3 is only for keys which start with `stress`. Another reason to switch to the new extxyz reader, which allows the explicit 2D syntax with `[[...]]`. MACE could do this directly rather than using ase.io, if you want to, as the extxyz is registered on `PyPI`. https://github.com/libAtoms/extxyz/

**Reply by bernstei** [reply]
True, and useful, but there may be enough legacy files that it'd still need to do the reshaping.

**Reply by jameskermode** [reply]
yes, some checking of array shapes and reshaping inside MACE is probably a sensible way to deal with this

**Reply by yonghyuklee** [reply]
Thank you very much! I also realized `virial` key works for the conversion from 9-element to 3x3 arrays.

**Reply by ilyes319** [reply]
I have just updated the code to allow for (9,), (3,3) and Voigt notation. Hopefully it will simplify all that.
